compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,188 +75,537 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
+        int32_t requiredBdIdsForInnerBatches =
+            dmaBatch->getRequiredBdIdsForInnerBatches();
+        dmaBatch->requiredBdIds =
+            requiredBdIdsForInnerBatches + totalDmaOpsInCurrentBatch;
       }
     }
   }
+};
+
+/// Create a BD ID generator for every shim tile.
+static void createShimTileToGeneratorMap(
+    AMDAIE::WorkgroupOp workgroupOp, AMDAIEDeviceModel &deviceModel,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap) {
+  workgroupOp->walk([&](AMDAIE::TileOp tileOp) {
+    std::optional<int64_t> col = getConstantIntValue(tileOp.getCol());
+    std::optional<int64_t> row = getConstantIntValue(tileOp.getRow());
+    if (col && row && deviceModel.isShimNOCTile(col.value(), row.value())) {
+      ChannelBdIdGenerator generator(
+          deviceModel.getChannelToValidBdIds(AMDAIETileType::SHIMNOC));
+      shimTileToGeneratorMap[tileOp.getResult()] = std::move(generator);
+    }
+  });
 }
 
-/// Creates a BD ID operation for the given DMA copy operation.
-/// If the DMA copy operation is inside a loop, the BD ID operation will be
-/// created with a semi-affine expression to assign different BD IDs for each
-/// iteration. Otherwise, a constant BD ID will be assigned.
-template <CopyOpOperateOn OperateOn>
-FailureOr<AMDAIE::BdIdOp> getBdIdOp(
-    IRRewriter &rewriter, AMDAIE::NpuDmaCpyNdOp &npuDmaOp,
-    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap,
-    DenseMap<AMDAIE::BdIdOp, SmallVector<uint32_t>> &bdIdOpToBdIdsMap) {
-  // Get the TileOp.
-  FailureOr<AMDAIE::TileOp> maybeTileOp =
-      getGeneratorTileOp<OperateOn>(npuDmaOp, shimTileToGeneratorMap);
-  if (failed(maybeTileOp)) return failure();
-  AMDAIE::TileOp tileOp = maybeTileOp.value();
-
-  // Get the channel.
-  FailureOr<AMDAIE::ChannelOp> maybeChannelOp;
-  if constexpr (OperateOn == CopyOpOperateOn::Source) {
-    maybeChannelOp = npuDmaOp.getSourceChannelOp();
-  } else if constexpr (OperateOn == CopyOpOperateOn::Target) {
-    maybeChannelOp = npuDmaOp.getTargetChannelOp();
-  } else {
-    return npuDmaOp.emitOpError()
-           << "Function can only operate on Source or Target";
+/// A struct that maintains the channel and source/target information for a
+/// given DmaOp and TileOp.
+struct DmaTileData {
+  uint32_t channel = -1;
+  uint32_t bdIdMapIndex = -1;
+};
+
+class BdIdAssignmentUtil {
+  // A mapping from a shim tile to its BD ID generator.
+  DenseMap<Value, ChannelBdIdGenerator> shimTileToGeneratorMap;
+  // A mapping from a BD ID Op to the BD IDs it covers.
+  DenseMap<AMDAIE::BdIdOp, SmallVector<uint32_t>> bdIdOpToBdIdsMap;
+  // A mapping from DMAOp to its corresponding source/target BD IDs.
+  DenseMap<AMDAIE::NpuDmaCpyNdOp, SmallVector<AMDAIE::BdIdOp, 2>>
+      dmaOpToBdIdMap;
+
+ public:
+  BdIdAssignmentUtil(
+      DenseMap<Value, ChannelBdIdGenerator> shimTileToGeneratorMap)
+      : shimTileToGeneratorMap(shimTileToGeneratorMap) {}
```

Comment:
```suggestion
      : shimTileToGeneratorMap(std::move(shimTileToGeneratorMap)) {}
```
This should avoid a copy.